### DISCUSSION TITLE
A RAG-System using Milvus

### DISCUSSION THREAD

**Discussion by TomTolleson** [milvus-io/milvus#38466] [Show and tell]
Hello! I'm just getting my feet wet with Milvus and as part of my self-training, I built a Retrieval-Augmented Generative system.

https://github.com/TomTolleson/RAG-System

Not sure if I should learn Milvus by going in a new direction or building on this project. If anyone has advice on that it would be great! If anyone wants to contribute or provide me with feedback on my RAG System project, that would be also very much appreciated.

Thanks!

THT

**Comment by yhmo** [comment]
Did you get any error? Seems no problem in your source code.

**Reply by TomTolleson** [reply]
Nope. It's all checking out fine here and in my GitHub CI. This was my first time using Codecov. I'm only getting 43% which I need to look into a way to improve that (if possible).
Hoping this can be useful to people wanting to create a domain-specific knowledge base.

**Reply by yhmo** [reply]
Basically, Milvus is a tool that finds similar embeddings by computing distances/similarity between embeddings. Milvus provides indexes and configurations to affect the recall rate. The text splitter and embedding model determine the embeddings, which is fundamental of recall. You can try using a different splitter or embedding model. For example, use RecursiveCharacterTextSplitter instead of CharacterTextSplitter.

**Reply by TomTolleson** [reply]
An, ok. So experimenting with the recursive text splitter could be a possible direction to go in as far as introducing a change?
If I do that I'd like to compare benchmarks on the default text splitter vs the recursive. Any recommendations on areas of focus for that?

Thanks!

THT

**Reply by yhmo** [reply]
Yes, this line will return different documents with different splitters:
documents = load_documents(str(data_path))
Then the OpenAI embedding model will generate different embeddings into milvus.
To use different embedding models, you can read the langchain document for more info.

The Milvus.from_documents() method eventually runs into langchain.Milvus.from_texts()
https://github.com/langchain-ai/langchain-milvus/blob/1d2c1abb3a55db7b855e9caf8482dbe2f627ba63/libs/milvus/langchain_milvus/vectorstores/milvus.py#L1454

The langchain.Milvus.from_texts() has a parameter "drop_old", which means to drop the old collection in milvus. The default value is False, which means it will reuse the old collection. In your program, if an old collection is reused, the previous embeddings might lead to unexpected results since they were generated by different splitters or models. So, you can set this parameter to True so that it creates a new collection each time in your experiment.

```
class MilvusStore:
    def add_documents(self, documents: List[Document]) -> Milvus:
        self.vector_store = Milvus.from_documents(
            documents,
            self.embeddings,
            connection_args={"host": MILVUS_HOST, "port": MILVUS_PORT},
            drop_old=True
        )
        return self.vector_store
```
